roiextractors/extractors/nwbextractors/nwbextractors.py

```diff
@@ -80,55 +116,112 @@ class NwbImagingExtractor(ImagingExtractor):
-        if start_frame is None:
-            start_frame = 0
-        if end_frame is None:
-            end_frame = self.get_num_frames()
-        end_frame = min(end_frame, self.get_num_frames())
-
-        video = self._video[start_frame: end_frame]
 
+        if len(opts.data.shape) == 3:
+            self._num_frames, self._size_x, self._size_y = opts.data.shape
+            self._num_channels = 1
+            self._channel_names = opts.imaging_plane.optical_channel[0].name
+        else:
+            raise NotImplementedError("4D volumetric data are currently not supported")
+
+            # Fill epochs dictionary
+            self._epochs = {}
+        if self.nwbfile.epochs is not None:
+            df_epochs = self.nwbfile.epochs.to_dataframe()
+            # TODO implement add_epoch() method in base class
+            self._epochs = {row['tags'][0]: {
+                'start_frame': self.time_to_frame(row['start_time']),
+                'end_frame': self.time_to_frame(row['stop_time'])}
+                for _, row in df_epochs.iterrows()}
+
+            self._kwargs = {'file_path': str(Path(file_path).absolute()),
+                            'optical_series_name': optical_series_name}
+        self.make_nwb_metadata(nwbfile=self.nwbfile, opts=opts)
+
+    def __del__(self):
+        self.io.close()
+
+    def time_to_frame(self, time: FloatType):
+        return int((time - self._imaging_start_time) * self.get_sampling_frequency())
+
+    def frame_to_time(self, frame: IntType):
+        return float(frame / self.get_sampling_frequency() + self._imaging_start_time)
+
+    def make_nwb_metadata(self, nwbfile, opts):
+        # Metadata dictionary - useful for constructing a nwb file
+        self.nwb_metadata = dict()
+        self.nwb_metadata['NWBFile'] = {
+            'session_description': nwbfile.session_description,
+            'identifier': nwbfile.identifier,
+            'session_start_time': nwbfile.session_start_time,
+            'institution': nwbfile.institution,
+            'lab': nwbfile.lab
+        }
+        self.nwb_metadata['Ophys'] = dict()
+        # Update metadata with Device info
+        self.nwb_metadata['Ophys']['Device'] = []
+        for dev in nwbfile.devices:
+            self.nwb_metadata['Ophys']['Device'].append({'name': dev})
+
+        # Update metadata with ElectricalSeries info
+        self.nwb_metadata['Ophys']['TwoPhotonSeries'] = []
+        self.nwb_metadata['Ophys']['TwoPhotonSeries'].append({
+            'name': opts.name
+        })
+
+    #TODO use lazy_ops
+    @check_get_frames_args
+    def get_frames(self, frame_idxs, channel=0):
+        opts = self.nwbfile.acquisition[self._optical_series_name]
+        if frame_idxs.size > 1 and np.all(np.diff(frame_idxs) > 0):
+            return opts.data[frame_idxs]
+        else:
+            sorted_idxs = np.sort(frame_idxs)
+            argsorted_idxs = np.argsort(frame_idxs)
+            return opts.data[sorted_idxs][argsorted_idxs]
```

Comment:
channel does not do anything here

Reply:
true, but it's in the signature of the base class method